subprojects/docs/src/docs/release/notes.md

```diff
@@ -69,6 +69,26 @@ ADD RELEASE FEATURES ABOVE
 
 ## New features and usability improvements
 
+### Improved incremental compilation for Java
+
+The Java incremental compiler received substantial improvements in this release:
+
+1. incremental compilation analysis is now stored in the build cache
+2. incremental compilation analysis is faster, uses less memory and disk space
+3. incremental compiler is now aware of constant consumers
+
+In previous Gradle releases, incremental compilation analysis was only stored locally.
+This means that if the compile task outputs were fetched from the build cache, then no incremental compilation was performed on the next build: the initial build after a fetch from the cache was fast, but the next one was always a full recompilation.
+In Gradle 7.1 this is no longer the case, as the result of the analysis is now stored in the build cache, meaning that the first compilation after fetching from the build cache will be incremental.
+
+The next improvement is about performance: incremental compilation analysis adds overhead to compilation, because Gradle needs to extract symbols from class files, analyze a transitive graph of dependencies to determine the consumers of a particular symbol, etc.
+In Gradle 7.1, we have significantly reduced this overhead, as well as reduced the cost of storage of the analysis.
+On the `gradle` project itself, we were able to make incremental compilation up to twice as fast!
+
+Last, because of the way the Java compiler works, previous Gradle releases were forced to perform a full recompilation as soon as _any_ constant was changed in an upstream dependency.
```

Comment:
For this change, we also need to update https://github.com/gradle/gradle/blob/6e7ba25c80638c3afff960f8923d861fd4ca88fb/subprojects/docs/src/docs/userguide/jvm/java_plugin.adoc#incremental-java-compilation

It also seems that the limitation in `Known issues` about "If there is a mismatch in the package declaration and the directory structure of source files" has been already fixed, so we can remove it as well, right?

And maybe we should also remove the part about constants in the incremental Groovy compiler section: https://github.com/gradle/gradle/blob/6e7ba25c80638c3afff960f8923d861fd4ca88fb/subprojects/docs/src/docs/userguide/jvm/groovy_plugin.adoc#incremental-groovy-compilation

Reply:
Fair point, let me do this.